lib/elixir/lib/calendar/duration.ex

```diff
@@ -0,0 +1,181 @@
+defmodule Duration do
+  @moduledoc """
+  Struct and functions for handling durations.
+
+  A `Duration` struct represents a collection of time scale units,
+  allowing for manipulation and calculation of durations.
+
+  Date and time scale units are represented as integers, allowing for both positive and negative values.
+
+  Microseconds are represented using a tuple `{microsecond, precision}`.
+  This ensures compatibility with other calendar types implementing time, such as `Time`, `DateTime`, and `NaiveDateTime`.
+  """
+
+  defstruct year: 0,
+            month: 0,
+            week: 0,
+            day: 0,
+            hour: 0,
+            minute: 0,
+            second: 0,
+            microsecond: {0, 0}
+
+  @type t :: %Duration{
+          year: integer,
+          month: integer,
+          week: integer,
+          day: integer,
+          hour: integer,
+          minute: integer,
+          second: integer,
+          microsecond: {integer, integer}
+        }
+
+  @type unit ::
+          {:year, integer}
+          | {:month, integer}
+          | {:week, integer}
+          | {:day, integer}
+          | {:hour, integer}
+          | {:minute, integer}
+          | {:second, integer}
+          | {:microsecond, {integer, integer}}
+
+  @doc """
+  Creates a new `Duration` struct from given `units`.
+
+  Raises a KeyError when called with invalid units.
+
+  ## Examples
+
+      iex> Duration.new(month: 2)
+      %Duration{month: 2}
+
+  """
+  @spec new([unit]) :: t
+  def new(units) do
```

Comment:
I think it is fine because anything matching the typespec should succeed, but should we consider making this `new!/1` and have a `new/1` returning a tuple?
Would also be symmetrical with other calendar structs like `Date.new!`.
(Sorry if I missed a previous discussion about this)

Reply:
All good! This topic was touched [here](https://github.com/elixir-lang/elixir/pull/13385/files/f8844a458c552fc57d606a02a1084d22f80af042#r1513939780) and [here](https://github.com/elixir-lang/elixir/pull/13385#discussion_r1544288939)

The TLDR being that we want to raise in `new/1` since it'll only do so because of invalid input coming from the user. There are no semantics involved like in `Date.new` where the input may be handled differently depending on the calendar that is used.

Reply:
I actually thought about something else. Should it be called `new!` for consistency? All of the calendar types return a tuple from `new`. Same for `URI.new`. But note we are not consistent with this in Elixir itself (Range.new, MapSet.new, Map.new do not return tuples).

Reply:
> But note we are not consistent with this in Elixir itself (Range.new, MapSet.new, Map.new do not return tuples).

True, but to be fair these are all primitive data types for collections.
With such a logic, consistency with calendar structs and URI might be more sensible here?

Reply:
Hmm i am a bit torn, i liked the reasoning behind "it's fine to raise due to erroneous user input", but i am all for consistency of course! I'd be fine to have `new/1` and `new!/1`, but if we were to use the non-bang version for `shift/2-3` in the calendar types we might be forced to return tuples from shift as well?

Reply:
@tfiedlerdejanze we can start with only `Duration.new!` for now and revisit later accordingly. But I think we always want `shift` to fail for a bad duration (it is invalid input after all).

Reply:
Fair enough, one more question: `new/1` currently raise a KeyError if provided with invalid keys for a `%Duration{}` and raises an `ArgumentError` if provided with invalid values for a duration unit... Should we keep that behaviour or raise an `ArgumentError` in either case instead?

Reply:
@tfiedlerdejanze I think raising any of them is fine. If it is easy and cheap to validate so we always raise `ArgumentError`, that works too.

Reply:
Actually since we're already validating all tuples, we could probably just add an `in` guard like `defp validate_duration_unit!({unit, value}) when is_integer(value) and unit in [:year, ...]` and then raise an `ArgumentError` in the last clause, should remain quite efficient?

I think a `KeyError` feels slightly weird, even if I don't really mind.

Reply:
You're right, it would! now that this was merged, we can follow up with that in a small extra PR 👌

Reply:
Yep, I will send a PR directly, Jose gave me a heads up before merging 👌